external-crates/move/documentation/book/src/index-syntax.md

```diff
@@ -0,0 +1,297 @@
+operations. It takes a mutable reference of the subject type and returns an mutable reference to
+the element type. The `borrow_mut` function defined in `std::vector` is an example of this:
+
+```move
+#[syntax(index)]
+public fun borrow_mut<Element>(v: &mut vector<Element>, i: u64): &mut Element {
+    // implementation
+}
+```
+
+#### Visibility
+
+To ensure that indexing functions are available anywhere the type is used, all index syntax methods
+must have public visibility. This ensures ergonomic usage of indexing across modules and packages in
+Move.
+
+#### No Duplicates
+
+In addition to the above requirements, we restrict each subject base type to defining a single
+index syntax method for immutable references and a single index syntax method for mutable
+references. For example, you may not define a specialized version for a polymorphic type:
+
+```move
+#[syntax(index)]
+public fun borrow_matrix_u64(s: &Matrix<u64>, i: u64, j: u64): &u64 { ... }
+
+#[syntax(index)]
+public fun borrow_matrix<T>(s: &Matrix<T>, i: u64, j: u64): &T { ... }
+    // ERROR! Matrix already has a definition
+    // for its immutable index syntax method
+```
+
+This ensures that the user can always tell which method is being invoked, without the need to
+inspect type instantiation.
+
+### Type Constraints
+
+By default, an index syntax method has the following type constraints:
+
+**Its subject type (first argument) must be a reference to a single type defined in the same module
+as the marked function.** This means that we cannot define index syntax methods for tuples,
+type parameters, or values:
+
+```move
+#[syntax(index)]
+public fun borrow_fst(x: &(u64, u64), ...): &u64 { ... }
+    // ERROR because the subject type is a tuple
+
+#[syntax(index)]
+public fun borrow_tyarg<T>(x: &T, ...): &T { ... }
+    // ERROR because the subject type is a type parameter
+
+#[syntax(index)]
+public fun borrow_value(x: Matrix<u64>, ...): &u64 { ... }
+    // ERROR because x is not a reference
+```
+
+**The subject type must match mutability with the return type.** This restriction allows users to
+clarify the expected behavior when they borrow an indexed expression as `&vec[i]` versus `&mut
+vec[i]`. The Move compiler uses the mutability marker to determine which borrow form to call to
+produce a reference of the appropriate mutability. As a result, we disallow index syntax methods
+whose subject and return mutability differ:
+
+```move
+#[syntax(index)]
+public fun borrow_imm(x: &mut Matrix<u64>, ...): &u64 { ... }
+    // ERROR! incompatible mutability
+
+```
+
+### Type Compatibility
+
+When defining an immutable and mutable index syntax method pair, they are subject to a number of
+compatibility constraints:
+
+1. They must take the same number of type parameters, those type parameters must have the same
+   constraints.
+5. Type parameters must be used the same _by position_, not name.
```

Comment:
I'd suggest just using 1s for ordered lists. Reader might get thrown off by the numbering sequence not being in order when list items get moved.